src/stdio.rs

```diff
@@ -0,0 +1,243 @@
+use super::Error;
+
+use core::mem::ManuallyDrop;
+
+use core::pin::Pin;
+use core::task::{Context, Poll};
+
+use std::fs::File;
+use std::io::{self, IoSlice};
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
+
+use std::process;
+
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, ReadBuf};
+
+#[derive(Debug)]
+pub(crate) enum StdioImpl {
+    /// Read/Write to /dev/null
+    Null,
+    /// Read/Write to a newly created pipe
+    Pipe,
+    /// Read/Write to custom fd
+    Fd(File),
+}
+
+/// Similar to std::process::Stdio
+#[derive(Debug)]
+pub struct Stdio(pub(crate) StdioImpl);
+impl Stdio {
+    /// Create a pipe for child communication
+    pub const fn piped() -> Self {
+        Self(StdioImpl::Pipe)
+    }
+
+    /// Pass /dev/null for child communication
+    pub const fn null() -> Self {
+        Self(StdioImpl::Null)
+    }
+}
+impl<T: IntoRawFd> From<T> for Stdio {
+    fn from(val: T) -> Self {
+        unsafe { Stdio::from_raw_fd(val.into_raw_fd()) }
+    }
+}
+impl FromRawFd for Stdio {
+    unsafe fn from_raw_fd(fd: RawFd) -> Self {
+        Self(StdioImpl::Fd(File::from_raw_fd(fd)))
+    }
+}
+impl From<Stdio> for process::Stdio {
+    fn from(stdio: Stdio) -> Self {
+        match stdio.0 {
+            StdioImpl::Null => process::Stdio::null(),
+            StdioImpl::Pipe => process::Stdio::piped(),
+            StdioImpl::Fd(fd) => unsafe {
+                process::Stdio::from_raw_fd(ManuallyDrop::new(fd).as_raw_fd())
+            },
+        }
+    }
+}
+
+#[derive(Debug)]
+enum ChildStdinImp {
+    ProcessImpl(tokio::process::ChildStdin),
+
+    #[cfg(feature = "mux_client")]
+    MuxClientImpl(super::mux_client_impl::ChildStdin),
+}
+
+/// Input for the remote child.
+#[derive(Debug)]
+pub struct ChildStdin(ChildStdinImp);
+
+impl From<tokio::process::ChildStdin> for ChildStdin {
+    fn from(imp: tokio::process::ChildStdin) -> Self {
+        Self(ChildStdinImp::ProcessImpl(imp))
+    }
+}
+
+#[cfg(feature = "mux_client")]
+impl From<super::mux_client_impl::ChildStdin> for ChildStdin {
+    fn from(imp: super::mux_client_impl::ChildStdin) -> Self {
+        Self(ChildStdinImp::MuxClientImpl(imp))
+    }
+}
+
+impl AsRawFd for ChildStdin {
+    fn as_raw_fd(&self) -> RawFd {
+        match &self.0 {
+            ChildStdinImp::ProcessImpl(imp) => AsRawFd::as_raw_fd(imp),
+
+            #[cfg(feature = "mux_client")]
+            ChildStdinImp::MuxClientImpl(imp) => AsRawFd::as_raw_fd(imp),
+        }
+    }
+}
+
+impl AsyncWrite for ChildStdin {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let inner = unsafe { Pin::into_inner_unchecked(self) };
+        match &mut inner.0 {
+            ChildStdinImp::ProcessImpl(imp) => {
+                AsyncWrite::poll_write(unsafe { Pin::new_unchecked(imp) }, cx, buf)
```

Comment:
Here too we should use `pin-project-lite`. The `Pin` invariants are _really_ easy to get wrong.

Reply:
Alternatively, in this case, I _think_ you can just use `Pin::new` (with no `unsafe`) since these are `Unpin`.

Reply:
I think we can use `pin-project` here since `openssh_mux_client` pulls in `serde`, which in turn pulls in `proc-macro`.

According to `pin-project-lite`:
> This is the only reason to use this crate. However, if you already have proc-macro related dependencies in your crate’s dependency graph, there is no benefit from using this crate. (Note: There is almost no difference in the amount of code generated between pin-project and pin-project-lite.)

Reply:
OK, with `pin-project`, I was able to remove all `unsafe` blocks.